synapse/storage/databases/main/appservice.py

```diff
@@ -203,19 +203,29 @@ async def get_appservice_state(
         """Get the application service state.
 
         Args:
-            service: The service whose state to set.
+            service: The service whose state to get.
         Returns:
-            An ApplicationServiceState or none.
+            An ApplicationServiceState, or None if we have yet to attempt any
+            transactions to the AS.
         """
-        result = await self.db_pool.simple_select_one(
+        # if we have created transactions for this AS but not yet attempted to send
+        # them, we will have a row in the table with state=NULL (recording the stream
+        # positions we have processed up to).
+        #
+        # On the other hand, if we have yet to create any transactions for this AS at
+        # all, then there will be no row for the AS.
+        #
+        # In either case, we return None to indicate "we don't yet know the state of
+        # this AS".
+        result = await self.db_pool.simple_select_one_onecol(
             "application_services_state",
             {"as_id": service.id},
-            ["state"],
+            retcol="state",
             allow_none=True,
             desc="get_appservice_state",
         )
         if result:
-            return ApplicationServiceState(result.get("state"))
+            return ApplicationServiceState(result)
```

Comment:
this changes things so that if there is a row for the AS but it has `state IS NULL`, we return `None` rather than raising an exception.

And I *think* this is necessary because fixing the race in  `set_appservice_stream_type_pos` now means that we hit this code at a point where we have a row with `{stream_type}_stream_id` but not yet a `state`. Previously, there would have (incorrectly) been no row at all.

Reply:
> this changes things so that if there is a row for the AS but it has `state IS NULL`, we return `None` rather than raising an exception.

It took me a second to understand that `ApplicationServiceState` is a enum, and so raises if given a `None`.

We appear to have changed this behaviour https://github.com/matrix-org/synapse/pull/11488, and previously we returned `None` if `state` was `None`.